### PR TITLE
[NET-661] Handle UDP Packets with missing addresses

### PR DESCRIPTION
**Pull request by mbakht**
### Use a new `kprobe`

For tracking UDP sends, earlier we were attaching a `kprobe` to 

```C
int udp_sendmsg(struct sock *sk, struct msghdr *msg, size_t len);
```

We were extracting _source_ and _destination_ IPs and ports from `sk`.

Unfortunately, it turned out that `sk` does not **always** have IP and port information. The socket has the destination address only if the socket was _connected_ at some point. (please see https://blog.packagecloud.io/eng/2017/02/06/monitoring-tuning-linux-networking-stack-sending-data/#udp_sendmsg for more details).

The `msg` struct (the second param passed to `sendmsg`) does have destination address and port in the case of _unconnected_ sockets. But source IP and port still remains missing.

That is why we have decided to attach `kprobe` to the following function instead:

```C
struct sk_buff *ip_make_skb(struct sock *sk, struct flowi4 *fl4,
			    int getfrag(void *from, char *to, int offset,
					int len, int odd, struct sk_buff *skb),
			    void *from, int length, int transhdrlen,
			    struct ipcm_cookie *ipc, struct rtable **rtp,
			    struct inet_cork *cork, unsigned int flags);
```

As we can see, the first parameter here is still `struct sock`. It allows us to reuse existing code for the _connected_ sockets to extract IPs and ports. In the case of _unconnected_ sockets, we now have a fallback option - `struct flowi4 *fl4`. This data structure has all the IP and port information.

Unfortunately, `struct flowi4` is *not portable* across kernel versions. That is where _offset guessing_ comes in.

### Offset guessing

We have already been using offset guessing to determine offsets in `struct sock`. We use the same approach for guessing offsets in `flowi4`. 

One difference is that all the earlier offset guessing was done using TCP connections. But `ip_make_skb` is specifically used for UDP connections. So the event generator had to be modified to generate UDP events.

#### Notes

- For IPv6, `ip6_make_skb` gets called instead of `ip_make_skb`. 

```C
struct sk_buff *ip6_make_skb(struct sock *sk,
			     int getfrag(void *from, char *to, int offset,
					 int len, int odd, struct sk_buff *skb),
			     void *from, int length, int transhdrlen,
			     struct ipcm6_cookie *ipc6, struct flowi6 *fl6,
			     struct rt6_info *rt, unsigned int flags,
			     const struct sockcm_cookie *sockc);
```
To handle _unconnected_ sockets, we will have to use offset guessing for `struct flowi6`. This is NOT part of this PR. It will be done as part of future work.

- We had following blocks of repeated code for each field we were guessing offsets for:

```go
			status.what = guessDaddrIPv6				
			logSuccessfulGuess(guessRTT, status.offset_rtt)	
			logStartGuess(guessDaddrIPv6)
````
These three lines have been factored to a common function: `logAndAdvance(status, status.offset_rtt, guessDaddrIPv6)`.

- We now need to extract IP and port from `conn` objects in two places. So the code has been factored out to a single function `extractIPsAndPorts`.

#### Unit tests:

Two new tests have been added
- One test checks for _unconnected_ UDP connections. This was particularly tricky since it was not clear how data can be sent to a _unconnected_ UDP socket using the net API of GoLang. 
- The other test checks that we can process IPv6 UDP packets when the socket is _connected_. 

#### Test on the agent VM

- Start the `system-probe`
- Do `dig abc.com @8.8.8.8`
- `curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections | python parser.py`
- the output looks something like this:

```
$ stats | jq . | python parser.py
2437  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
2426  udp    127.0.0.1:37047 ->       127.0.0.1:8125   0  0  0  0  0
2381  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
2457  udp    10.0.2.15:36161 ->         8.8.8.8:53     1  0  0 46591  0
2453  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
7223  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
2451  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
2407  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
2456  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
2457  udp    127.0.0.1:44091 ->       127.0.0.1:44091  0  0  0  0  0
12878 tcp    10.0.2.15:22    ->        10.0.2.2:60677  0  0  0  0  0
2454  udp 192.168.33.11:873   ->    192.168.33.1:2049   0  0  0  0  0
```

Notice that we are now seeing information for `8.8.8.8:53 `. Also, we are also seeing `192.168.33.1:2049` (NFS) connections which were not seen before.

The `parser.py` is a handy script to spit out connections in an easy to read way:

```python
import json
import sys

d = json.load(sys.stdin)
for c in d['conns']:
    s =  "{: <5} {} {: >12}:{: <5} ->{: >16}:{: <5} {: >2} {: >2} {: >2} {: >2} {: >2}"
    print s.format(
        c['pid'],
        c['type'],
        c['laddr']['ip'],
        c['laddr']['port'],
        c['raddr']['ip'],
        c['raddr']['port'],
        c['dnsSuccessfulResponses'],
        c['dnsFailedResponses'],
        c['dnsTimeouts'],
        c['dnsSuccessLatencySum'],
        c['dnsFailureLatencySum'],
    )
```

### CODE REVIEW

File: pkg/ebpf/c/tracer-ebpf.c [hunk 1]
```diff
@@ -634,19 +659,61 @@ int kretprobe__tcp_close(struct pt_regs* ctx) {
...
+     }
+
+     log_debug("kprobe/ip6_make_skb: pid_tgid: %d, size: %d\n", pid_tgid, size);
+     handle_message(&t, size, 0);
+     increment_telemetry_count(udp_send_processed);
+
+     return 0;
+}
+
+SEC("kprobe/ip_make_skb")
```
Comment (p-lambert) [inline pkg/ebpf/c/tracer-ebpf.c:684]: curious, could we use this kprobe for both TCP and UDP? it seems that this would triggered for both, right?

File: pkg/ebpf/tracer_test.go [hunk 2]
```diff
@@ -1732,3 +1731,52 @@ func TestTCPEstablishedPreExistingConn(t *testing.T) {
 	assert.Equal(t, uint32(0), conn.MonotonicTCPEstablished)
 	assert.Equal(t, uint32(1), conn.MonotonicTCPClosed)
 }
+
+func TestUnconnectedUDPSendIPv4(t *testing.T) {
```
Comment (p-lambert) [inline pkg/ebpf/tracer_test.go:1735]: 👍

File: pkg/ebpf/c/tracer-ebpf.c (hunk 1 above)
Comment (mbakht) [reply pkg/ebpf/c/tracer-ebpf.c:684]: As far as I know, it works only for UDP.
From. https://elixir.bootlin.com/linux/latest/ident/ip_make_skb:

Screen Shot 2020-09-04 at 9 56 03 AM

Comment (p-lambert) [reply pkg/ebpf/c/tracer-ebpf.c:684]: Gotcha. Another thing worth checking is if the arguments of this function haven't changed since kernel 4.4.0

Comment (mbakht) [reply pkg/ebpf/c/tracer-ebpf.c:684]: My understanding so far is that unlike UDP, TCP is not limited to one method for creating SKB. It calls different functions depending on the context (e.g., SKBs for retransmission are treated different than a regular transmission).

File: pkg/ebpf/c/tracer-ebpf.c [hunk 3]
```diff
@@ -284,6 +308,12 @@ static __always_inline int read_conn_tuple(conn_tuple_t* t, struct sock* skp, u6
     t->pid = pid_tgid >> 32;
     t->metadata = type;
 
+    // Retrieve network namespace id first since addresses and ports may not be available for unconnected UDP
+    // sends
```
[the remark this answers is not in the document]
Comment (mbakht) [reply pkg/ebpf/c/tracer-ebpf.c:312]: No - it doesn't. But when address/ports are unavailable, we exit this function _before_ extracting the network namespace id. But then later in the `ip_make_skb`, we might be able to get address/ports. That means, the network id would remain unavailable unless we put this code block there too.

**Comment by mbakht** [comment]
Two things changed in the latest update:
- After talking with @p-lambert, removed special check for RTT offsets since it has been in production for quite a while
- Added special checks for offsets for `flowi4` going over `threshold`. If it does, instead of terminating the whole offset guessing process, we simply set a `boolean` flag (`fl4_offsets`) to tell `tracer-ebpf.c` that we don't know the `flowi4` offsets.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 4]
```diff
@@ -634,19 +665,61 @@ int kretprobe__tcp_close(struct pt_regs* ctx) {
...
 
+    size = size - sizeof(struct udphdr);
+
     conn_tuple_t t = {};
     if (!read_conn_tuple(&t, sk, pid_tgid, CONN_TYPE_UDP)) {
-        increment_telemetry_count(udp_send_missed);
-        return 0;
+        struct flowi4* fl4 = (struct flowi4*)PT_REGS_PARM2(ctx);
+        bpf_probe_read(&t.saddr_l, sizeof(u32), ((char*)fl4) + offset_saddr_fl4());
+        bpf_probe_read(&t.daddr_l, sizeof(u32), ((char*)fl4) + offset_daddr_fl4());
```
Comment (ISauve) [inline pkg/ebpf/c/tracer-ebpf.c:702]: Out of curiosity, why is the size `u32` instead of `u64`? Are we truncating `t.saddr_l` and `t.daddr_l` for a specific reason here?

**Review by ISauve — APPROVED** [review]
LGTM, though you'll probably want someone else to take a look as well :)

File: pkg/ebpf/c/tracer-ebpf.c (hunk 1 above)
Comment (leeavital) [reply pkg/ebpf/c/tracer-ebpf.c:690]: > could we use this kprobe for both TCP and UDP

It might be worth adding a comment somewhere that this is only used in the UDP code path. I remember the name of the function causing me to ask the same question.

File: pkg/ebpf/tracer_test.go [hunk 5]
```diff
@@ -1732,3 +1731,52 @@ func TestTCPEstablishedPreExistingConn(t *testing.T) {
...
+	cfg := NewDefaultConfig()
+	tr, err := NewTracer(cfg)
+	require.NoError(t, err)
+	defer tr.Stop()
+
+	remotePort := rand.Int()%5000 + 15000
+	remoteAddr := &net.UDPAddr{IP: net.ParseIP("8.8.8.8"), Port: remotePort}
+	// Use ListenUDP instead of DialUDP to create a "connectionless" UDP connection
+	conn, err := net.ListenUDP("udp", nil)
```
Comment (leeavital) [inline pkg/ebpf/tracer_test.go:1744]: clever!

**Review by leeavital — APPROVED** [review]

File: pkg/ebpf/c/tracer-ebpf.c (hunk 4 above)
Comment (brycekahle) [reply pkg/ebpf/c/tracer-ebpf.c:702]: IPv4 addresses are only 32 bits.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 6]
```diff
@@ -634,19 +665,61 @@ int kretprobe__tcp_close(struct pt_regs* ctx) {
...
+
+        if (!t.saddr_l || !t.daddr_l) {
+            log_debug("ERR(fl4): src/dst addr not set src:%d,dst:%d\n", t.saddr_l, t.daddr_l);
+            increment_telemetry_count(udp_send_missed);
+            return 0;
+        }
+        bpf_probe_read(&t.sport, sizeof(t.sport), ((char*)fl4) + offset_sport_fl4());
+        bpf_probe_read(&t.dport, sizeof(t.dport), ((char*)fl4) + offset_dport_fl4());
+
+         if (t.sport == 0 || t.dport == 0) {
```
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.c:712]: nit: there is some weird indentation going on here. Can you check and make it consistent? (I hope to introduce a formatter soon).

File: pkg/ebpf/offsetguess.go [hunk 7]
```diff
@@ -106,46 +122,64 @@ type fieldValues struct {
...
 }
 
-func expectedValues(conn net.Conn) (*fieldValues, error) {
-	netns, err := ownNetNS()
+func extractIPsAndPorts(conn net.Conn) (
+	saddr, daddr uint32,
+	sport, dport uint16,
+	err error,
+) {
+	saddr_s, sport_s, err := net.SplitHostPort(conn.LocalAddr().String())
```
Comment (brycekahle) [inline pkg/ebpf/offsetguess.go:138]: nit: underscores are not something we usually use for variable names in Go. Camelcasing is preferred, e.g. `saddrStr`

File: pkg/ebpf/c/tracer-ebpf.h [hunk 8]
```diff
@@ -3,8 +3,8 @@
 
 #include <linux/types.h>
 
-static const __u64 TRACER_IPV6_DISABLED = 0;
-static const __u64 TRACER_IPV6_ENABLED = 1;
+static const __u64 ENABLED = 0;
+static const __u64 DISABLED = 1;
```
Comment (brycekahle) [inline pkg/ebpf/c/tracer-ebpf.h:7]: You have switched the values of enabled and disabled here. Is that intentional?

File: pkg/ebpf/bytecode/types.go [hunk 9]
```diff
@@ -112,7 +114,7 @@ var (
 	// the actual eBPF function that it should bind to
 	KProbeOverrides = map[ProbeName]ProbeName{
 		TCPSendMsgPre410: TCPSendMsg,
-		UDPSendMsgPre410: UDPSendMsg,
+		UDPSendMsgPre410: IPMakeSkb,
```
[the remark this answers is not in the document]
Comment (brycekahle) [reply pkg/ebpf/bytecode/types.go:117]: I just checked 3.10 (which is what CentOS 7.6 uses) and it is the same, so we shouldn't need this anymore.

**Comment by brycekahle** [comment]
Since UDP is by nature connectionless, I find the terminology of "connected" and "unconnected" confusing. Could we clarify exactly what is meant by those terms?

**Comment by brycekahle** [comment]
@mbakht It looks like you made a commit from the agent dev VM, which is causing the CLA bot to complain about a commit from `vagrant`. You might need to squash or rebase and fix that commit. See https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit for some ways to do so.

File: pkg/ebpf/tracer_test.go [hunk 10]
```diff
@@ -1547,7 +1547,6 @@ const (
 func testDNSStats(t *testing.T, domain string, success int, failure int, timeout int, serverIP string) {
 	config := NewDefaultConfig()
 	config.CollectDNSStats = true
-	config.CollectLocalDNS = true
```
Comment (brycekahle) [inline pkg/ebpf/tracer_test.go:1550]: Why was this removed?

File: pkg/ebpf/c/tracer-ebpf.h (hunk 8 above)
Comment (mbakht) [reply pkg/ebpf/c/tracer-ebpf.h:7]: No, it is not intentional. Great catch! I will check why it was not being caught by any of the tests.

Comment (brycekahle) [reply pkg/ebpf/c/tracer-ebpf.h:7]: 🙇‍♂️  I was wondering the same thing about the tests.

File: pkg/ebpf/tracer_test.go (hunk 10 above)
Comment (mbakht) [reply pkg/ebpf/tracer_test.go:1516]: This was added before unnecessarily since none of the tracked connections are local.

**Comment by mbakht** [comment]
> @mbakht It looks like you made a commit from the agent dev VM, which is causing the CLA bot to complain about a commit from vagrant. You might need to squash or rebase and fix that commit. See https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit for some ways to do so.

You are right - I did mistakenly made a commit from the agent VM 🤦 . I tried to follow the instructions in the link you shared. Unfortunately, because I have already merged `master` into this branch, the suggestion mentioned there did not work. I tried a bunch of other approaches - one of them worked but messed up the commit hashes. If merging remains blocked because of this, I will create a new branch and PR with this code once you approve.

> Since UDP is by nature connectionless, I find the terminology of "connected" and "unconnected" confusing. Could we clarify exactly what is meant by those terms?

I agree that it is indeed confusing in the context of UDP. Interestingly enough, even in kernel it uses comparison against a TCP state to figure out if an UDP connection is "connected" or not: https://elixir.bootlin.com/linux/latest/source/net/ipv4/udp.c#L1038

**Comment by mbakht** [comment]
@brycekahle , addressed all the comments. Please have another look when you get a chance.

**Review by brycekahle — APPROVED** [review]
